ambari-server/src/main/java/org/apache/ambari/server/checks/UpgradeCheckRegistry.java

```diff
@@ -17,165 +17,185 @@
  */
 package org.apache.ambari.server.checks;
 
-import java.io.File;
-import java.io.FilenameFilter;
-import java.net.URL;
-import java.net.URLClassLoader;
 import java.util.ArrayList;
 import java.util.Comparator;
+import java.util.HashMap;
+import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.TreeSet;
 
 import org.apache.ambari.annotations.UpgradeCheckInfo;
+import org.apache.ambari.server.AmbariException;
+import org.apache.ambari.server.api.services.AmbariMetaInfo;
 import org.apache.ambari.server.stack.upgrade.UpgradePack;
-import org.apache.ambari.server.state.ServiceInfo;
+import org.apache.ambari.server.state.CheckHelper;
+import org.apache.ambari.server.state.StackId;
+import org.apache.ambari.server.state.StackInfo;
+import org.apache.ambari.spi.upgrade.CheckQualification;
 import org.apache.ambari.spi.upgrade.UpgradeCheck;
 import org.apache.ambari.spi.upgrade.UpgradeCheckGroup;
 import org.apache.ambari.spi.upgrade.UpgradeType;
+import org.apache.commons.lang.ArrayUtils;
+import org.apache.commons.lang.StringUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import org.springframework.util.ClassUtils;
 
-import com.google.inject.Singleton;
+import com.google.inject.Inject;
+import com.google.inject.Injector;
+import com.google.inject.Provider;
 
 /**
  * The {@link UpgradeCheckRegistry} contains the ordered list of all pre-upgrade
  * checks. This will order the checks according to
  * {@link PreUpgradeCheckComparator}.
+ * <p>
+ * There are two types of checks which can be loaded into this registry:
+ * <ul>
+ * <li>Built-in: checks which ship with Ambari
+ * <li>Plug-in: checks which are specified in the upgrade pack for a stack, and
+ * either come from Ambari's checks or from the library classpath of the stack.
+ * </ul>
  */
-@Singleton
 public class UpgradeCheckRegistry {
   private static final Logger LOG = LoggerFactory.getLogger(UpgradeCheckRegistry.class);
 
+  @Inject
+  private Provider<AmbariMetaInfo> metainfoProvider;
+
   /**
-   * The list of upgrade checks to run through.
+   * Used for injecting dependencies into plugin upgrade check classes. There
+   * are cases where an upgrade pack specifies a built-in check which uses a
+   * dependency, such as the Ambari Configuration singleton.
    */
-  private Set<UpgradeCheck> m_upgradeChecks = new TreeSet<>(
+  @Inject
+  private Injector m_injector;
+
+  /**
+   * The list of upgrade checks provided by the system which are required, as
+   * specified by {@link UpgradeCheckInfo#required()}.
```

Comment:
Doesn't `m_builtInChecks` have all built-in checks?

Reply:
Yes, it will contain all built-in checks.